core/src/main/java/org/opensearch/remote/metadata/common/SdkClientUtils.java

```diff
@@ -9,25 +9,178 @@
+     */
+    public static BiConsumer<DeleteDataObjectResponse, Throwable> wrapDeleteCompletion(ActionListener<DeleteResponse> listener) {
+        return (r, throwable) -> {
+            if (throwable == null) {
+                try {
+                    DeleteResponse deleteResponse = DeleteResponse.fromXContent(r.parser());
+                    listener.onResponse(deleteResponse);
+                } catch (IOException e) {
+                    handleParseFailure(listener, "delete");
+                }
+            } else {
+                handleThrowable(listener, throwable);
+            }
+        };
+    }
+
+    /**
+     * Wraps the completion of a BULK operation from the SdkClient into a format compatible with an ActionListener.
+     *
+     * @param listener The ActionListener that will receive the parsed BulkResponse or any errors
+     * @return A BiConsumer that can be used directly with CompletionStage's whenComplete method
+     */
+    public static BiConsumer<BulkDataObjectResponse, Throwable> wrapBulkCompletion(ActionListener<BulkResponse> listener) {
+        return (r, throwable) -> {
+            if (throwable == null) {
+                try {
+                    BulkResponse bulkResponse = BulkResponse.fromXContent(r.parser());
+                    listener.onResponse(bulkResponse);
+                } catch (IOException e) {
+                    handleParseFailure(listener, "bulk");
+                }
+            } else {
+                handleThrowable(listener, throwable);
+            }
+        };
+    }
+
+    /**
+     * Wraps the completion of a SEARCH operation from the SdkClient into a format compatible with an ActionListener.
+     *
+     * @param listener The ActionListener that will receive the parsed SearchResponse or any errors
+     * @return A BiConsumer that can be used directly with CompletionStage's whenComplete method
+     */
+    public static BiConsumer<SearchDataObjectResponse, Throwable> wrapSearchCompletion(ActionListener<SearchResponse> listener) {
+        return (r, throwable) -> {
+            if (throwable == null) {
+                try {
+                    SearchResponse searchResponse = SearchResponse.fromXContent(r.parser());
+                    listener.onResponse(searchResponse);
+                } catch (IOException e) {
+                    handleParseFailure(listener, "search");
+                }
+            } else {
+                handleThrowable(listener, throwable);
+            }
+        };
+    }
+
+    private static void handleParseFailure(ActionListener<?> listener, String operation) {
+        listener.onFailure(new OpenSearchStatusException("Failed to parse " + operation + " response", INTERNAL_SERVER_ERROR));
+    }
+
+    private static void handleThrowable(ActionListener<?> listener, Throwable throwable) {
+        Exception exception = unwrapAndConvertToException(throwable, CompletionException.class, OpenSearchStatusException.class);
```

Comment:
Will that be possible for users to provide any exception class too?

Reply:
Not with this util: it's for convenience only. How would you suggest making this more flexible?

Reply:
may be `wrapGetCompletion` can take another parameter like `Class<? extends Throwable>... exceptionTypesToUnwrap`?